File: erpnext/accounts/report/tax_detail/tax_detail.js
// Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
// Contributed by Case Solved and sponsored by Nulight Studios
/* eslint-disable */

frappe.provide('frappe.query_reports');

frappe.query_reports["Tax Detail"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("company"),
			reqd: 1
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.month_start(frappe.datetime.get_today()),
			reqd: 1,
			width: "60px"
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.month_end(frappe.datetime.get_today()),
			reqd: 1,
			width: "60px"
		},
		{
			fieldname: "report_name",
			label: __("Report Name"),
			fieldtype: "Read Only",
			default: frappe.query_report.report_name,
			hidden: 1,
			reqd: 1
		},
		{
			fieldname: "mode",
			label: __("Mode"),
			fieldtype: "Read Only",
			default: "edit",
			hidden: 1,
			reqd: 1
		}
	],
	onload: function onload(report) {
		// Remove Add Column and Save from menu
		report.page.add_inner_button(__("New Report"), () => new_report(), __("Custom Report"));
		report.page.add_inner_button(__("Load Report"), () => load_report(), __("Custom Report"));
		hide_filters(report);
	}
};

function hide_filters(report) {
	report.page.page_form[0].querySelectorAll('.form-group.frappe-control').forEach(function setHidden(field) {
		if (field.dataset.fieldtype == "Read Only") {
			field.classList.add("hidden");
		}
	});
}

erpnext.TaxDetail = class TaxDetail {
	constructor() {
		this.patch();
		this.load_report();
	}
	// Monkey patch the QueryReport class
	patch() {
		this.qr = frappe.query_report;
		this.super = {
			refresh_report: this.qr.refresh_report,
			show_footer_message: this.qr.show_footer_message
		}
		this.qr.refresh_report = () => this.refresh_report();
		this.qr.show_footer_message = () => this.show_footer_message();
	}
	show_footer_message() {
		// The last thing to run after datatable_render in refresh()
		this.super.show_footer_message.apply(this.qr);
		if (this.qr.report_name !== 'Tax Detail') {
			this.show_help();
			if (this.loading) {
				this.set_section('');
			} else {
				this.reload_component('');
			}
		}
		this.loading = false;
	}
	refresh_report() {
		// Infrequent report build (onload), load filters & data
		// super function runs a refresh() serially
		// already run within frappe.run_serially
		this.loading = true;
		this.super.refresh_report.apply(this.qr);
		if (this.qr.report_name !== 'Tax Detail') {
			frappe.call({
				method: 'erpnext.accounts.report.tax_detail.tax_detail.get_custom_reports',
				args: {name: this.qr.report_name}
			}).then((r) => {
				const data = JSON.parse(r.message[this.qr.report_name]['json']);
				this.create_controls();
				this.sections = data.sections || {};
				this.controls['show_detail'].set_input(data.show_detail);
			});
		}
	}
	load_report() {
		// One-off report build like titles, menu, etc
		// Run when this object is created which happens in qr.load_report
		this.qr.menu_items = this.get_menu_items();
	}
	get_menu_items() {
		// Replace Save action
		let new_items = [];
		const save = __('Save');

		for (let item of this.qr.menu_items) {
			if (item.label === save) {
				new_items.push({
					label: save,
					action: () => this.save_report(),
					standard: false
				});
			} else {
				new_items.push(item);
			}
		}
		return new_items;
	}
	save_report() {
		this.check_datatable();
		if (this.qr.report_name !== 'Tax Detail') {
			frappe.call({
				method:'erpnext.accounts.report.tax_detail.tax_detail.save_custom_report',
				args: {
					reference_report: 'Tax Detail',
					report_name: this.qr.report_name,
					data: {
						columns: this.qr.get_visible_columns(),
						sections: this.sections,
						show_detail: this.controls['show_detail'].get_input_value()
					}
				},
				freeze: true
			}).then((r) => {
				this.set_section('');
			});
		}
	}
	check_datatable() {
		if (!this.qr.datatable) {
			frappe.throw(__('Please change the date range to load data first'));
		}
	}
	set_section(name) {
		// Sets the given section name and then reloads the data
		if (name && !this.sections[name]) {
			this.sections[name] = {};
		}
		let options = Object.keys(this.sections);
		options.unshift('');
		this.controls['section_name'].$wrapper.find("select").empty().add_options(options);
		const org_mode = this.qr.get_filter_value('mode');
		let refresh = false;
		if (name) {
			this.controls['section_name'].set_input(name);
			this.qr.set_filter_value('mode', 'edit');
			if (org_mode === 'run') {
				refresh = true;
			}
		} else {
			this.controls['section_name'].set_input('');
			this.qr.set_filter_value('mode', 'run');
			if (org_mode === 'edit') {
				refresh = true;
			}
		}
		if (refresh) {
			this.qr.refresh();
		}
		this.reload_component('');
	}
	reload_component(component_name) {
		const section_name = this.controls['section_name'].get_input_value();
		if (section_name) {
			const section = this.sections[section_name];
			const component_names = Object.keys(section);
			component_names.unshift('');
			this.controls['component'].$wrapper.find("select").empty().add_options(component_names);
			this.controls['component'].set_input(component_name);
			if (component_name) {
				this.controls['component_type'].set_input(section[component_name].type);
			}
		} else {
			this.controls['component'].$wrapper.find("select").empty();
			this.controls['component'].set_input('');
		}
		this.set_table_filters();
	}
	set_table_filters() {
		let filters = {};
		const section_name = this.controls['section_name'].get_input_value();
		const component_name = this.controls['component'].get_input_value();
		if (section_name && component_name) {
			const component_type = this.sections[section_name][component_name].type;
			if (component_type === 'filter') {
				filters = this.sections[section_name][component_name]['filters'];
			}
		}
		this.setAppliedFilters(filters);
	}
	setAppliedFilters(filters) {
		if (this.qr.datatable) {
			Array.from(this.qr.datatable.header.querySelectorAll('.dt-filter')).map(function setFilters(input) {
				let idx = input.dataset.colIndex;
				if (filters[idx]) {
					input.value = filters[idx];
				} else {
					input.value = null;
				}
			});
			this.qr.datatable.columnmanager.applyFilter(filters);
		}
	}
	delete(name, type) {
		if (type === 'section') {
			delete this.sections[name];
			const new_section = Object.keys(this.sections)[0] || '';
			this.set_section(new_section);
		}
		if (type === 'component') {
			const cur_section = this.controls['section_name'].get_input_value();
			delete this.sections[cur_section][name];
			this.reload_component('');
		}
	}
	create_controls() {
		let controls = {};
		// SELECT in data.js
		controls['section_name'] = this.qr.page.add_field({
			label: __('Section'),
			fieldtype: 'Select',
			fieldname: 'section_name',
			change: (e) => {
				this.set_section(this.controls['section_name'].get_input_value());
			}
		});
		// BUTTON in button.js
		controls['new_section'] = this.qr.page.add_field({
			label: __('New Section'),
			fieldtype: 'Button',
			fieldname: 'new_section',
			click: () => {
				frappe.prompt({
					label: __('Section Name'),
					fieldname: 'name',
					fieldtype: 'Data'
				}, (values) => {
					this.set_section(values.name);
				});
			}
		});
		controls['delete_section'] = this.qr.page.add_field({
			label: __('Delete Section'),
			fieldtype: 'Button',
			fieldname: 'delete_section',
			click: () => {
				let cur_section = this.controls['section_name'].get_input_value();
				if (cur_section) {
					frappe.confirm(__('Are you sure you want to delete section') + ' ' + cur_section + '?',
					() => {this.delete(cur_section, 'section')});
				}
			}
		});
		controls['component'] = this.qr.page.add_field({
			label: __('Component'),
			fieldtype: 'Select',
			fieldname: 'component',
			change: (e) => {
				this.reload_component(this.controls['component'].get_input_value());
			}
		});
		controls['component_type'] = this.qr.page.add_field({
			label: __('Component Type'),
			fieldtype: 'Select',
			fieldname: 'component_type',
			default: 'filter',
			options: [
				{label: __('Filtered Row Subtotal'), value: 'filter'},
				{label: __('Section Subtotal'), value: 'section'}
			]
		});
		controls['add_component'] = this.qr.page.add_field({
			label: __('Add Component'),
			fieldtype: 'Button',
			fieldname: 'add_component',
			click: () => {
				this.check_datatable();
				let section_name = this.controls['section_name'].get_input_value();
				if (section_name) {
					const component_type = this.controls['component_type'].get_input_value();
					let idx = 0;
					const names = Object.keys(this.sections[section_name]);
					if (names.length > 0) {
						const idxs = names.map((key) => parseInt(key.match(/\d+$/)) || 0);
						idx = Math.max(...idxs) + 1;
					}
					const filters = this.qr.datatable.columnmanager.getAppliedFilters();
					if (component_type === 'filter') {
						const name = 'Filter' + idx.toString();
						let data = {
							type: component_type,
							filters: filters
						}
						this.sections[section_name][name] = data;
						this.reload_component(name);
					} else if (component_type === 'section') {
						if (filters && Object.keys(filters).length !== 0) {
							frappe.show_alert({
								message: __('Column filters ignored'),
								indicator: 'yellow'
							});
						}
						let data = {
							type: component_type
						}
						frappe.prompt({
							label: __('Section'),
							fieldname: 'section',
							fieldtype: 'Select',
							options: Object.keys(this.sections)
						}, (values) => {
							this.sections[section_name][values.section] = data;
							this.reload_component(values.section);
						});
					} else {
						frappe.throw(__('Please select the Component Type first'));
					}
				} else {
					frappe.throw(__('Please select the Section first'));
				}
			}
		});
		controls['delete_component'] = this.qr.page.add_field({
			label: __('Delete Component'),
			fieldtype: 'Button',
			fieldname: 'delete_component',
			click: () => {
				const component = this.controls['component'].get_input_value();
				if (component) {
					frappe.confirm(__('Are you sure you want to delete component') + ' ' + component + '?',
					() => {this.delete(component, 'component')});
				}
			}
		});
		controls['save'] = this.qr.page.add_field({
			label: __('Save & Run'),
			fieldtype: 'Button',
			fieldname: 'save',
			click: () => {
				this.save_report();
			}
		});
		controls['show_detail'] = this.qr.page.add_field({
			label: __('Show Detail'),
			fieldtype: 'Check',
			fieldname: 'show_detail',
			default: 1
		});
		this.controls = controls;
	}
	show_help() {
		const help = __('Your custom report is built from General Ledger Entries within the date range. You can add multiple sections to the report using the New Section button. Each component added to a section adds a subset of the data into the specified section. Beware of duplicated data rows. The Filtered Row component type saves the datatable column filters to specify the added data. The Section component type refers to the data in a previously defined section, but it cannot refer to its parent section. The Amount column is summed to give the section subtotal. Use the Show Detail box to see the data rows included in each section in the final report. Once finished, hit Save & Run. Report contributed by');
		this.qr.$report_footer.append('<div class="col-md-12"><strong>' + __('Help') + `: </strong>${help}<a href="https://www.casesolved.co.uk"> Case Solved</a></div>`);
	}
}

if (!window.taxdetail) {
	window.taxdetail = new erpnext.TaxDetail();
}

function get_reports(cb) {
	frappe.call({
		method: 'erpnext.accounts.report.tax_detail.tax_detail.get_custom_reports',
		freeze: true
	}).then((r) => {
		cb(r.message);
	})
}

function new_report() {
	const dialog = new frappe.ui.Dialog({
		title: __('New Report'),
		fields: [
			{
				fieldname: 'report_name',
				label: __('Report Name'),
				fieldtype: 'Data',
				default: 'VAT Return'
			}
		],
		primary_action_label: __('Create'),
		primary_action: function new_report_pa(values) {
			frappe.call({
				method:'erpnext.accounts.report.tax_detail.tax_detail.save_custom_report',
				args: {
					reference_report: 'Tax Detail',
					report_name: values.report_name,
					data: {
						columns: [],
						sections: {},
						show_detail: 1
					}
				},
				freeze: true
			}).then((r) => {
				frappe.set_route('query-report', values.report_name);
			});
			dialog.hide();
		}
	});
	dialog.show();
}

function load_report() {
	get_reports(function load_report_cb(reports) {
		const dialog = new frappe.ui.Dialog({
			title: __('Load Report'),
			fields: [
				{
					fieldname: 'report_name',
					label: __('Report Name'),
					fieldtype: 'Select',
					options: Object.keys(reports)
				}
			],
			primary_action_label: __('Load'),
			primary_action: function load_report_pa(values) {
				dialog.hide();
				frappe.set_route('query-report', values.report_name);
			}
		});
		dialog.show();
	});
}

File: erpnext/accounts/report/tax_detail/tax_detail.py
# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt
# Contributed by Case Solved and sponsored by Nulight Studios


import json

import frappe
from frappe import _

# NOTE: Payroll is implemented using Journal Entries which are included as GL Entries

# field lists in multiple doctypes will be coalesced
required_sql_fields = {
	("GL Entry", 1): ["posting_date"],
	("Account",): ["root_type", "account_type"],
	("GL Entry", 2): ["account", "voucher_type", "voucher_no", "debit", "credit"],
	("Purchase Invoice Item", "Sales Invoice Item"): [
		"base_net_amount",
		"item_tax_rate",
		"item_tax_template",
		"item_group",
		"item_name",
	],
	("Purchase Invoice", "Sales Invoice"): ["taxes_and_charges", "tax_category"],
}


def execute(filters=None):
	if not filters:
		return [], []

	fieldlist = required_sql_fields
	fieldstr = get_fieldstr(fieldlist)

	gl_entries = frappe.db.sql(
		"""
		select {fieldstr}
		from `tabGL Entry` ge
		inner join `tabAccount` a on
			ge.account=a.name and ge.company=a.company
		left join `tabSales Invoice` si on
			ge.company=si.company and ge.voucher_type='Sales Invoice' and ge.voucher_no=si.name
		left join `tabSales Invoice Item` sii on
			a.root_type='Income' and si.name=sii.parent
		left join `tabPurchase Invoice` pi on
			ge.company=pi.company and ge.voucher_type='Purchase Invoice' and ge.voucher_no=pi.name
		left join `tabPurchase Invoice Item` pii on
			a.root_type='Expense' and pi.name=pii.parent
		where
			ge.company=%(company)s and
			ge.posting_date>=%(from_date)s and
			ge.posting_date<=%(to_date)s
		order by ge.posting_date, ge.voucher_no
		""".format(
			fieldstr=fieldstr
		),
		filters,
		as_dict=1,
	)

	report_data = modify_report_data(gl_entries)
	summary = None
	if filters["mode"] == "run" and filters["report_name"] != "Tax Detail":
		report_data, summary = run_report(filters["report_name"], report_data)

	# return columns, data, message, chart, report_summary
	return get_columns(fieldlist), report_data, None, None, summary


def run_report(report_name, data):
	"Applies the sections and filters saved in the custom report"
	report_config = json.loads(frappe.get_doc("Report", report_name).json)
	# Columns indexed from 1 wrt colno
	columns = report_config.get("columns")
	sections = report_config.get("sections", {})
	show_detail = report_config.get("show_detail", 1)
	report = {}
	new_data = []
	summary = []
	for section_name, section in sections.items():
		report[section_name] = {"rows": [], "subtotal": 0.0}
		for component_name, component in section.items():
			if component["type"] == "filter":
				for row in data:
					matched = True
					for colno, filter_string in component["filters"].items():
						filter_field = columns[int(colno) - 1]["fieldname"]
						if not filter_match(row[filter_field], filter_string):
							matched = False
							break
					if matched:
						report[section_name]["rows"] += [row]
						report[section_name]["subtotal"] += row["amount"]
			if component["type"] == "section":
				if component_name == section_name:
					frappe.throw(_("A report component cannot refer to its parent section") + ": " + section_name)
				try:
					report[section_name]["rows"] += report[component_name]["rows"]
					report[section_name]["subtotal"] += report[component_name]["subtotal"]
				except KeyError:
					frappe.throw(
						_("A report component can only refer to an earlier section") + ": " + section_name
					)

		if show_detail:
			new_data += report[section_name]["rows"]
		new_data += [{"voucher_no": section_name, "amount": report[section_name]["subtotal"]}]
		summary += [
			{"label": section_name, "datatype": "Currency", "value": report[section_name]["subtotal"]}
		]
		if show_detail:
			new_data += [{}]
	return new_data or data, summary or None


def filter_match(value, string):
	"Approximation to datatable filters"
	import datetime

	if string == "":
		return True
	if value is None:
		value = -999999999999999
	elif isinstance(value, datetime.date):
		return True

	if isinstance(value, str):
		value = value.lower()
		string = string.lower()
		if string[0] == "<":
			return True if string[1:].strip() else False
		elif string[0] == ">":
			return False if string[1:].strip() else True
		elif string[0] == "=":
			return string[1:] in value if string[1:] else False
		elif string[0:2] == "!=":
			return string[2:] not in value
		elif len(string.split(":")) == 2:
			pre, post = string.split(":")
			return True if not pre.strip() and post.strip() in value else False
		else:
			return string in value
	else:
		if string[0] in ["<", ">", "="]:
			operator = string[0]
			if operator == "=":
				operator = "=="
			string = string[1:].strip()
		elif string[0:2] == "!=":
			operator = "!="
			string = string[2:].strip()
		elif len(string.split(":")) == 2:
			pre, post = string.split(":")
			try:
				return True if float(pre) <= value and float(post) >= value else False
			except ValueError:
				return False if pre.strip() else True
		else:
			return string in str(value)

	try:
		num = float(string) if string.strip() else 0
		return frappe.safe_eval(f"{value} {operator} {num}")
	except ValueError:
		if operator == "<":
			return True
		return False


def abbrev(dt):
	return "".join(l[0].lower() for l in dt.split(" ")) + "."


def doclist(dt, dfs):
	return [abbrev(dt) + f for f in dfs]


def as_split(fields):
	for field in fields:
		split = field.split(" as ")
		yield (split[0], split[1] if len(split) > 1 else split[0])


def coalesce(doctypes, fields):
	coalesce = []
	for name, new_name in as_split(fields):
		sharedfields = ", ".join(abbrev(dt) + name for dt in doctypes)
		coalesce += [f"coalesce({sharedfields}) as {new_name}"]
	return coalesce


def get_fieldstr(fieldlist):
	fields = []
	for doctypes, docfields in fieldlist.items():
		if len(doctypes) == 1 or isinstance(doctypes[1], int):
			fields += doclist(doctypes[0], docfields)
		else:
			fields += coalesce(doctypes, docfields)
	return ", ".join(fields)


def get_columns(fieldlist):
	columns = {}
	for doctypes, docfields in fieldlist.items():
		fieldmap = {name: new_name for name, new_name in as_split(docfields)}
		for doctype in doctypes:
			if isinstance(doctype, int):
				break
			meta = frappe.get_meta(doctype)
			# get column field metadata from the db
			fieldmeta = {}
			for field in meta.get("fields"):
				if field.fieldname in fieldmap.keys():
					new_name = fieldmap[field.fieldname]
					fieldmeta[new_name] = {
						"label": _(field.label),
						"fieldname": new_name,
						"fieldtype": field.fieldtype,
						"options": field.options,
					}
			# edit the columns to match the modified data
			for field in fieldmap.values():
				col = modify_report_columns(doctype, field, fieldmeta[field])
				if col:
					columns[col["fieldname"]] = col
	# use of a dict ensures duplicate columns are removed
	return list(columns.values())


def modify_report_columns(doctype, field, column):
	"Because data is rearranged into other columns"
	if doctype in ["Sales Invoice Item", "Purchase Invoice Item"]:
		if field in ["item_tax_rate", "base_net_amount"]:
			return None

	if doctype == "GL Entry" and field in ["debit", "credit"]:
		column.update({"label": _("Amount"), "fieldname": "amount"})

	if field == "taxes_and_charges":
		column.update({"label": _("Taxes and Charges Template")})
	return column


def modify_report_data(data):
	import json

	new_data = []
	for line in data:
		if line.debit:
			line.amount = -line.debit
		else:
			line.amount = line.credit
		# Remove Invoice GL Tax Entries and generate Tax entries from the invoice lines
		if "Invoice" in line.voucher_type:
			if line.account_type not in ("Tax", "Round Off"):
				new_data += [line]
				if line.item_tax_rate:
					tax_rates = json.loads(line.item_tax_rate)
					for account, rate in tax_rates.items():
						tax_line = line.copy()
						tax_line.account_type = "Tax"
						tax_line.account = account
						if line.voucher_type == "Sales Invoice":
							line.amount = line.base_net_amount
							tax_line.amount = line.base_net_amount * (rate / 100)
						if line.voucher_type == "Purchase Invoice":
							line.amount = -line.base_net_amount
							tax_line.amount = -line.base_net_amount * (rate / 100)
						new_data += [tax_line]
		else:
			new_data += [line]
	return new_data


# JS client utilities

custom_report_dict = {
	"ref_doctype": "GL Entry",
	"report_type": "Custom Report",
	"reference_report": "Tax Detail",
}


@frappe.whitelist()
def get_custom_reports(name=None):
	filters = custom_report_dict.copy()
	if name:
		filters["name"] = name
	reports = frappe.get_list("Report", filters=filters, fields=["name", "json"], as_list=False)
	reports_dict = {rep.pop("name"): rep for rep in reports}
	# Prevent custom reports with the same name
	reports_dict["Tax Detail"] = {"json": None}
	return reports_dict


@frappe.whitelist()
def save_custom_report(reference_report, report_name, data):
	if reference_report != "Tax Detail":
		frappe.throw(_("The wrong report is referenced."))
	if report_name == "Tax Detail":
		frappe.throw(_("The parent report cannot be overwritten."))

	doc = {
		"doctype": "Report",
		"report_name": report_name,
		"is_standard": "No",
		"module": "Accounts",
		"json": data,
	}
	doc.update(custom_report_dict)

	try:
		newdoc = frappe.get_doc(doc)
		newdoc.insert()
		frappe.msgprint(_("Report created successfully"))
	except frappe.exceptions.DuplicateEntryError:
		dbdoc = frappe.get_doc("Report", report_name)
		dbdoc.update(doc)
		dbdoc.save()
		frappe.msgprint(_("Report updated successfully"))
	return report_name
